platform/pulp/download/batch.py

```diff
@@ -0,0 +1,527 @@
+
+
+class BatchExecutor(ThreadPoolExecutor):
+    """
+    Batch thread pool executor.
+    """
+
+    def __init__(self, concurrent=CONCURRENT, backlog=BACKLOG):
+        """
+        A thread pool executor tailored for the batch.
+        The worker queue size is restricted to limit memory footprint.
+
+        Args:
+            concurrent (int): The number of downloads to execute in concurrently.
+            backlog (int): The number of downloads kept in memory.
+
+        Raises:
+            ValueError: concurrent less than 2 or backlog is less than concurrent.
+        """
+        super(BatchExecutor, self).__init__(max_workers=concurrent)
+        self._work_queue = Queue(maxsize=backlog)
+        if concurrent < 2:
+            raise ValueError(_('concurrent may not be < 2'))
+        if backlog < concurrent:
+            raise ValueError(_('backlog may not be < concurrent'))
+
+
+class DownloadFeeder(Thread):
+    """
+    Download feeder.
+    A thread used to feed each batched download into the executor.
+    May be interrupted and terminated by calling shutdown().
+
+    Attributes:
+        batch (Batch): A batch to feed.
+        total (int): The total number of downloads queued.
+        is_shutdown (bool): Feeder has been shutdown.
+    """
+
+    def __init__(self, batch):
+        super(DownloadFeeder, self).__init__(name='feeder')
+        self.batch = batch
+        self.daemon = True
+        self.total = 0
+        self.is_shutdown = False
+
+    @property
+    def iterator(self):
+        return self.batch.iterator
+
+    @property
+    def executor(self):
+        return self.batch.executor
+
+    @property
+    def downloads(self):
+        return self.batch.downloads
+
+    @property
+    def context(self):
+        return self.batch.context
+
+    def shutdown(self, wait=True):
+        """
+        Shutdown.
+        Abort feeding and terminate.
+
+        Args:
+            wait (bool): Wait for thread to terminate.
+        """
+        if self.is_shutdown:
+            return
+        self.is_shutdown = True
+        if wait:
+            self.join()
+
+    def run(self):
+        """
+        Thread (main) loop.
+        Submit each download to the batch executor.
+        """
+        try:
+            for download in self.downloads:
+                if self.is_shutdown:
+                    return
+                log.debug(
+                    _('%(feeder)s - feed #%(total)d url=%(url)s'),
+                    {
+                        'feeder': self,
+                        'total': self.total,
+                        'url': download.url
+                    })
+                download.context = self.context
+                future = self.executor.submit(Plan(self.batch, download))
+                future.add_done_callback(self.iterator.add)
+                self.total += 1
+        except Exception as e:
+            log.exception(_('feeder failed.'))
+            self.iterator.raised(e)
```

Comment:
Your understanding is correct.  This logic is designed to cause an exception to be raised only when a component of the _framework_ itself fails (in this example, the feeder).  The result is - the caller, who is iterating the result of the batch _execution_, will be interrupted by a raised exception ASAP.